0005-kzen-two-party-ecdsa.adoc

```diff
@@ -0,0 +1,173 @@
+Some of them are in the above crate but some of them are imported from elsewhere.
+We'll start with going through these.
+
+== Key Generation
+
+=== Generating the joint public key
+
+To produce signatures under a joint public key you first need to generate it fairly.
+Because of rogue key attacks, you can't just have one party send their public key and the other send one back.
+
+The technique used in the paper (and followed by the implementation) is to do a coin tossing like protocol.
+This ensures that the public key is totally random and cannot be biased by the party that chooses their key afterwards.
+
+Note, The ability to simulate a random coin toss gives the key generation very strong security guarantees (they are able to do a simulation based proof).
+
+The protocol is roughly as follows:
+
+- P1 -> Hash(p1_public_key, schnorr_nizk(p1_public_key))
+- P2 -> p2_public_key, schnorr_nizk(p2_public_key)
+- P1 -> Opens commitment from first message
+- Both parties calculate `joint_public_key = p1_public_key + p2_public_key`.
+
+==== Implementation notes
+The implementation commits by sending two hashes, one for the key and one for the NIZK.
+
+The NIZK is implemented here: https://github.com/KZen-networks/curv/blob/18f0081a9a3025eafd49ad496e5032647edf4aa3/src/cryptographic_primitives/proofs/sigma_dlog.rs
+
+=== Proving Paillier modulus is well formed
+
+In order to use Paillier encryption P1 needs to generate a RSA type modulus `N = pq` where `p` and `q` are large primes.
+P1 needs to prove that she generated `N` properly (I don't actually know what the risk is to P2 if she doesn't).
+
+The proof mentioned in the original paper is <<Hazay11>>.
+But in a more recent paper on multi-party ECDSA, <<Lindell18>> modifies the proof from <<Goldberg18>> to do the job (see section 6.2.3).
+Roughly speaking, the strategy is to prove that `N` is *square free* and to show that `f(x) = x ^ N (mod N)` is a permutation which proves that `N = pq` form.
+
+==== Implementation notes
+
+Since it's a public coin protocol, you can make an non-interactive version of the proof using the [Fiat-Shamir transformation](https://en.wikipedia.org/wiki/Fiat%E2%80%93Shamir_heuristic).
+They've implemented a non-interactive version of it here: https://github.com/KZen-networks/zk-paillier/blob/a7f3907fb2b3464f12b55e3ad2f50405db54f36a/src/zkproofs/correct_key_ni.rs (which uses the interactive proof module but supplies the challenge as a hash of the initial commitment).
+
+
+=== Proving a Paillier encryption is a private key of a EC public key
+
+To do the signing protocol, we need one party (P2) to have the encrypted pivate key of the other (P1).
+In order for P2 to know that he's signing under their joint public key, he has to know that this encrypted private key is the right one (without learning what the private key is).
+
+The proof is described in <<Lindell17>> (section 6).
+It's a little tricky to get your head around.
+First P2 sends a challenge to P1.
```

Comment:
The challenge in this case means "sign this thing for me and I will verify against your PK"?

Reply:
No. It was probably a mistake to try and explain the protocol here. I expanded it so it at least provides some value but it's best to just try and read the paper.